keps/sig-instrumentation/3498-extending-stability/README.md

```diff
@@ -0,0 +1,179 @@
+
+- parsing variables
+- multi-line strings
+- evaluating buckets
+- buckets which are defined via variables and consts
+- evaluation of simple consts
+- evaluation of simple variables
+
+We will not attempt to parse metrics which:
+
+- are constructed dynamically, i.e. through function calls which use function arguments as parameters in metric definitions, since some of those cannot be resolved until runtime. 
+- are constructed using custom prometheus collectors, for the same reasons as above. 
+
+As an aside, much of this work has already been done, but is stashed in a local repo. 
+
+### Test Plan
+
+We have static analysis testing for stable metrics, we will extend our test coverage 
+to include metrics which are `ALPHA` and `BETA` while ignoring `INTERNAL` metrics.
+
+
+### Graduation Criteria
+
+
+#### Alpha
+
+- Kubernetes metrics framework will be enhanced to support additional stability classes
+- The static analysis pipeline of the metrics framework will be enhanced to understand how to parse more things (these are listed above)
+- All instances of `Alpha` metrics will be converted to `Internal`
+
+#### Beta
+
+- Kubernetes metrics framework will be enhanced to support marking `Alpha` and `Beta` metrics with a date. The semantics of this are yet to be determined. This date will be used to statically determine whether or not that metric should be decrepated automatically or promoted.
+- Kubernetes metrics framework will be enhanced with a script to auto-deprecate metrics which have passed their window of existence as an `Alpha` or `Beta` metric
+- It is at this point, we will determine the longevity rules for `Alpha` and `Beta` metrics
+- The `beta` stage for this framework will be a few releases. During this time, we will evaluate the utility and the ergonomics of the framework, making adjustments as necessary
+
+#### GA
+
+
+## Production Readiness Review Questionnaire
+
+During the `alpha` stage of this KEP, we will not be making any user facing changes, except marking metrics as `Internal` which were previously `Alpha`. The stability guarantees of `Internal` metrics is the same as `Alpha` currently and therefore there will not be any changes to what users can expect from the metrics they are using. 
+
+### Feature Enablement and Rollback
+
+We can revert our changes if it breaks the metrics framework. But we will be adding testing coverage as we enhance the framework, so it is unlikely that this will need to occur.
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+Metrics stability framework is an internal feature of Kubernetes. 
+
+###### Does enabling the feature change any default behavior?
+
+No.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, this can be rolled back.
```

Comment:
How?
In PRR we give instruction for operators - so they can't rollback the code. They need a knob to rollback (which I don't see here).

Reply:
That's true, it cannot be rolled back per-se, but the code can be rolled back upstream if there is an issue.